Datapath Plugins (CFP-41634) Part 5: Documentation And Example (#46872)
cilium-team/cilium: Documentation/network/ebpf/plugins.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. warning::
   This is a feature for advanced users with deep knowledge of Cilium and its
   architecture. Be very sure you actually need this before using it, as you
   can break Cilium's functionality or even compromise security. Use it at your
   own risk and know what you're doing.

.. warning::
    This is a beta feature. Please provide feedback and file a GitHub issue if

>     you experience any problems.

Reviewer: nit:

```suggestion
.. include:: ../beta.rst
```

Reviewer: For consistency, this inclusion of the beta.rst is also typically placed immediately _after_ the title of the feature that the beta message relates to.

Author: Done

################
Datapath Plugins
################

Datapath plugins let you inject your own BPF programs throughout the Cilium
datapath to influence datapath behavior or build custom observability tooling.

> Datapath plugins are deployed as a ``DaemonSets`` and talk to Cilium agents via

Reviewer: nit for consistency:
```suggestion
Datapath plugins are deployed as a ``DaemonSet`` and talk to Cilium agents via
```

Author: Done.

a gRPC interface. Datapath plugins are intended primarily as means to enable
datapath extension for operators with advanced needs and advanced knowledge of
Cilium's inner workings who are willing to continually maintain their plugin
as Cilium's internal architecture evolves. They are *not* intended to enable an
ecosystem of third party off-the-shelf plugins and there are *little to no*
compatibility guarantees for datapath plugins between Cilium versions. With

> that in mind, we will explore how datapath plugins work, what capabilities they

Reviewer: nit: voice
```suggestion
that in mind, this guide will explore how datapath plugins work, what capabilities they
```

Author: Done

provide, and how to build and deploy one.

Overview
========

.. image:: ../../_static/datapath-plugins-overview.png

Datapath plugins are deployed as a ``DaemonSet``. An instance of the plugin sits
on each node talking to the local Cilium agent via gRPC over a Unix domain
socket.

As Cilium is loading a BPF collection and preparing it for an attachment point
or set of related attachment points, whether they be TC attachments for a
network interface, an XDP attachment, or a cgroup attachment as with SOCK_ADDR

> and SOCKET hooks, it sends two rounds of gRPC requests to all registered

Reviewer: nit: This would be easier to read if the sentence skipped the `whether they be ... (different kinds)` and just communicated one idea first about how it works overall. Then as a second sentence "You can attach plugins at TC, XDP, cgroup, ...".

Author: Rewrote this to be:

> As Cilium is loading a BPF collection and preparing it for an attachment point
or set of related attachment points it sends two rounds of gRPC requests to all
registered plugins allowing them to instrument the collection with hooks that
run before (pre) or after (post) programs in that collection. Plugins can attach
hooks around TC, XDP, or cgroup SOCK_ADDR/SOCKET programs.

plugins allowing them to instrument the collection with hooks that run before
(pre) or after (post) programs in that collection. For now, only *entrypoints*
are allowed to be instrumented, the top-level program Cilium attaches to a BPF
attachment point, but this constraint may be loosened in the future to allow

> for more fine-grained instrumentation of the collection by a plugin.

Reviewer: nit: Shorter, clear sentences
```suggestion
(pre) or after (post) programs in that collection. For now, only *entrypoints*
are allowed to be instrumented, which are the top-level programs Cilium attaches 
to a BPF attachment point. This constraint may be loosened in the future to allow
for more fine-grained instrumentation of the collection by a plugin.
```

Author: Done

   are actually created at this point.
5. Cilium sends an ``InstrumentCollection`` request to any plugin that provided
   a hook specification in the prepare phase. Each ``InstrumentCollection``
   request contains the same context about the collection and attachment point
   that was in the preceding ``PrepareCollection`` request along with a list of
   hooks mirroring those requested by the plugin in its ``PrepareCollection``
   response. Each hook contains an attach target providing all the context the
   plugin needs to load a ``BPF_PROG_TYPE_EXT`` hook program into the
   kernel, and the unique ``pin_path`` for this hook program that Cilium
   generated in the previous step.
6. Each plugin loads hook programs and pins them to the designated pin paths
   before responding to the ``InstrumentCollection`` request.
7. Cilium opens and unpins the hook programs then replaces the dispatcher
   subprograms with these programs creating a set of freplace links in their
   place.
8. With the collection finalized, Cilium attaches collection entrypoints
   to the relevant BPF attachment points in the kernel.

Enabling Datapath Plugins
=========================

> .. _enabling_datapath_plugins:

Reviewer: We typically place the label _before_ the relevant header.
```suggestion
.. _enabling_datapath_plugins:

Enabling Datapath Plugins
=========================
```

Author: Done

.. tabs::

    .. group-tab:: Helm

       .. cilium-helm-install::
          :namespace: kube-system
          :set: datapathPlugins.enabled=true

    .. group-tab:: ConfigMap

       .. code-block:: yaml

           enable-datapath-plugins: true

Plugin Registration
===================

To register a plugin with Cilium, create a ``CiliumDatapathPlugin`` resource:

.. code-block:: yaml

    apiVersion: cilium.io/v2alpha1
    kind: CiliumDatapathPlugin
    metadata:
      name: example
    spec:
      attachmentPolicy: Always
      version: 0.0.1

``attachmentPolicy`` can be ``Always`` or ``BestEffort``. ``Always`` means that
Cilium must be able to talk to the plugin to successfully load and attach any
BPF collection. ``BestEffort`` means that if Cilium cannot talk to the plugin,

> it can proceed without consulting the plugin. Any time a

Reviewer: nit: Readability could be clearer with an unordered list:

```suggestion
The ``attachmentPolicy`` must be one of the following values:
* ``Always``: Cilium must be able to talk to the plugin to successfully load and attach any BPF collection. Plugin failures will cause Cilium to fail to reload BPF programs.
* ``BestEffort``: If Cilium cannot talk to the plugin, Cilium will ignore the plugin and proceed to load BPF programs without the plugin. 

Any time a
```

I'm assuming it's a required field--if not, may be worth mentioning the behavior if the user doesn't explicitly configure it.

Author: Done, and yes it's required.

``CiliumDatapathPlugin`` is created, updated, or deleted Cilium triggers a full
datapath reinitialization. Bump the ``version`` string if you upgrade your
plugin and want Cilium to reinitialize; Cilium will not reinitialize the
datapath if your plugin's pod shuts down or is replaced.

Talking To Cilium
=================

> Plugins should create a subdirectory that matches the ``name`` of their

Reviewer: This paragraph says "should". Is this meant to be "must"? What are the consequences if a plugin fails to do so? I guess this depends on the `attachmentPolicy` but basically Cilium cannot connect if the path is wrong?

Author: Changed to "must". Yes, Cilium wouldn't be able to talk and the behavior would depend on the `attachmentPolicy`. The plugin health is something I want to expose through the CLI or possibly part of the `CiliumDatapathPlugin` `status` eventually.

> ``CiliumDatapathPlugin`` CR under the configured ``datapath-plugins-state-dir``

Reviewer: nit: Avoid undefined acronyms
```suggestion
``CiliumDatapathPlugin`` resource under the configured ``datapath-plugins-state-dir``
```

Author: Done

and listen on a Unix socket called ``plugin.sock``. For example, if your
plugin's name is "example" and ``datapath-plugins-state-dir`` is

> ``/var/run/cilium/plugins``, your plugin should  create a socket at

Reviewer: nit: These extra words don't really add clarity, probably we can simplify the wording here:
```suggestion
``/var/run/cilium/plugins``, create a socket at
```

Author: Done

collection. ``InstrumentCollection`` happens after the BPF collection is loaded
into the kernel. Cilium passes BPF collection details to the plugin along with
details about hook attachment points it created in the prepare phase. The plugin
loads its BPF programs and passes them back to Cilium to be attached to these
hook points. See the full gRPC API listing below for details on the structure
of the request and response messages.

.. code-block:: go

   func (s *datapathPluginServer) PrepareCollection(ctx context.Context, req *datapathplugins.PrepareCollectionRequest) (*datapathplugins.PrepareCollectionResponse, error) {
       // Determine which programs in the collection to instrument and tell
       // Cilium in the PrepareCollectionResponse.
   }

   func (s *datapathPluginServer) InstrumentCollection(ctx context.Context, req *datapathplugins.InstrumentCollectionRequest) (*datapathplugins.InstrumentCollectionResponse, error) {
       // Load and pin BPF programs for each hook point specified in the
       // request. The set of hooks will match those passed to Cilium in the
       // associated PrepareCollection response.
   }

CAP_SYS_ADMIN

> =============

Reviewer: nit:

```suggestion
Security Requirements
=====================
```

I wonder if it would be also helpful to have some clear description of the potential security impact and expectations around datapath plugins. Maybe the simplest approach would be to tell the reader that Cilium assumes that plugins are running with root-equivalent privileges and that it is the reader's responsibility to secure the plugins; failure to properly secure the plugins can lead to the cluster becoming compromised at a level similar to cluster admin.

Author: ```
Security Requirements
=====================

Cilium passes program and map IDs for the programs and maps in the collection
in the ``InstrumentCollection`` request. Depending on your plugin's needs, you
may wish to pass these maps or programs to your own BPF collection as you
load your hooks. In order to get a BPF object ID from a file descriptor, you
must grant ``SYS_ADMIN`` to your plugin's pod:

.. code-block:: yaml

   securityContext:
     capabilities:
       drop:
       - ALL
       add:
       # BPF ID -> FD translation requires CAP_SYS_ADMIN.
       - SYS_ADMIN

Such plugins run with root-equivalent privileges. It is the plugin developer's
responsibility to secure their plugin; a compromised plugin can result in a
compromised cluster at a level similar to cluster admin.
```

I updated it to this. I rephrased a little bit from your suggestion. Note that `SYS_ADMIN` isn't a hard requirement unless you need to do the ID->FD translation. 

Also, just thinking out loud... in the future it probably makes sense to have a way to do this that doesn't require `SYS_ADMIN`. We could expand the plugin interface so that the plugin tells Cilium in the prepare phase which maps from a collection it wants to use and Cilium includes a pin path for the map in the `InstrumentCollection` request.

Reviewer: Thanks, change LGTM.

(no change required) We can debate whether "root-equivalent" is quite the most accurate representation here, but I think it communicates at least the right message. I see this also as not only the privileges strictly required by the plugin to do the plugin's job here, but also that there is a risk of confused deputy. The plugin to some degree dictates what Cilium will do with Cilium's wider privileges. It is arguably a security improvement to remove the requirement for `CAP_SYS_ADMIN`. At the same time the Linux capabilities are not the only security permissions exposed here.

Cilium passes program and map IDs for the programs and maps in the collection
in the ``InstrumentCollection`` request. Depending on your plugin's needs, you
may wish to pass these maps or programs to your own BPF collection as you
load your hooks. In order to get a BPF object ID from a file descriptor, you
must grant ``SYS_ADMIN`` to your plugin's pod:

.. code-block:: yaml

   securityContext:
     capabilities:
       drop:
       - ALL
       add:
       # BPF ID -> FD translation requires CAP_SYS_ADMIN.
       - SYS_ADMIN

Putting It All Together: Example Plugin
=======================================

Cilium provides an `example plugin <https://github.com/cilium/cilium/examples/datapath-plugin>`_
that illustrates how to instrument each kind of collection and program. To
deploy it, make sure that datapath plugins :ref:`are enabled<enabling_datapath_plugins>`
in your kind cluster then run the following commands:

.. code-block:: bash

   $ make kind-image-datapath-plugin

>    $ make kind-install-datapath-plugin

Reviewer 2: Just noticed: this should actually be kind-install-cilium-datapath-plugin right?

Author: Yep. Fixed.
